pybamm_tea/tea.py

```diff
@@ -0,0 +1,690 @@
+            self.parameter_values.update(pava, check_already_exists=False)
+
+    @property
+    def stack_breakdown(self):
+        """A breakdown of volume- and mass- loadings on stack level."""
+        try:
+            return self._stack_breakdown
+        except AttributeError:
+            self._stack_breakdown = self.calculate_stack_breakdown()
+            return self._stack_breakdown
+
+    @property
+    def stack_breakdown_dataframe(self):
+        """
+        A dataframe with components, volume-, mass- loadings and densities on stack
+        level.
+        """
+        try:
+            return self._stack_breakdown_dataframe
+        except AttributeError:
+            self._stack_breakdown_dataframe = self.print_stack_breakdown()
+            return self._stack_breakdown_dataframe
+
+    @property
+    def stack_energy_densities(self):
+        """Energy densities and parameters for calculation on stack level."""
+        try:
+            return self._stack_energy_densities
+        except AttributeError:
+            self._stack_energy_densities = self.calculate_stack_energy_densities()
+            return self._stack_energy_densities
+
+    def print_stack_breakdown(self):
+        """
+        A dataframe with components, volume-, mass loadings and densities on stack
+        level.
+        """
+        stack_bd = self.stack_breakdown
+
+        # Create a dataframe with columns for components, volume- and mass loadings
+        # and densities
+        components = pd.DataFrame(
+            [
+                c.replace(" volume loading [uL.cm-2]", "")
+                for c in stack_bd.keys()
+                if " volume loading [uL.cm-2]" in c
+            ],
+            columns=["Components"],
+        )
+        volumes = pd.DataFrame(
+            [
+                stack_bd.get(f"{c} volume loading [uL.cm-2]")
+                for c in components["Components"]
+            ],
+            columns=["Volume loading [uL.cm-2]"],
+        )
+        masses = pd.DataFrame(
+            [
+                stack_bd.get(f"{c} mass loading [mg.cm-2]")
+                for c in components["Components"]
+            ],
+            columns=["Mass loading [mg.cm-2]"],
+        )
+        densities = pd.DataFrame(
+            [stack_bd.get(f"{c} density [mg.uL-1]") for c in components["Components"]],
+            columns=["Density [mg.uL-1]"],
+        )
+
+        # Concatenate the dataframes side by side
+        df = pd.concat([components, volumes, masses, densities], axis=1)
+        df.set_index("Components", inplace=True)
+        df.index.name = None
+        return df
+
+    def calculate_stack_energy_densities(self):
+        """
+        Calculate ideal volumetric and gravimetric energy densities on stack
+        level.
+        """
+        stack_ed = {}  # stack energy densities dict
+        pava = self.parameter_values  # parameter values
+
+        # stoichiometries at 0 and 100% SOC
+        try:
```

Comment:
don't use a bare try/except. catch the error that is raised and raise an informative error here